[Bug 19056] lcb: Add safe foreign handler support

runrevmark opened the pull request:
This patch allows a foreign handler definition to be prefixed by '__safe'
indicating that it is safe to call from safe context.

The motivation for this is to ease binding to C/C++ functions which are
written to be easily callable from LCB.

runrevmark committed:
lcb: Add safe foreign handler support

This patch allows a foreign handler definition to be prefixed by '__safe'
indicating that it is safe to call from safe context.

The motivation for this is to ease binding to C/C++ functions which are
written to be easily callable from LCB.

peter-b commented:
@runrevmark Currently, we are reserving lowercase ASCII words (i.e. any token that matches `[a-z]+`) for syntax and keywords.  Notably, `__[a-z]+` will _not_ trigger a warning if used as a variable name.  I suggest that we turn change the "not suitable for identifier" warning to match on `_*[a-z]+`.

runrevmark committed:
lcb: Amend reserved identifier pattern to _*[a-z]+

This patch amends the reserved identifier pattern (which currently
produces a warning) to be _*[a-z]+.

Additionally, it fixes the check for reserved identifiers which had
been broken by the namespace additions and adds a front-end compiler
test.

runrevmark commented:
@peter-b: Done!

peter-b commented on toolchain/lc-compile/src/check.g:
```diff
@@ -1754,8 +1754,6 @@
 		Id'Namespace -> Namespace
         (|
-			where(Namespace -> id(_))
-		||
```
@runrevmark, I think that in places where identifiers are qualified, then they don't count as reserved identifiers.  For example, when I write `com.livecode.canvas`, that shouldn't trigger a "reserved identifier" warning, but when I write `variable __naughty`, it should.  It looks like the builds are failing because the compiler is incorrectly emitting warnings for module names.

livecodeali replied:
Ah yes, now it's coming back to me that that's why I added that line...

runrevmark replied:
That would explain its presence...

So the aim is to warn people if they make a definition with name matching pattern _*[a-z]+. This means we only need to check definition id's (e.g. variable <id>) not all ids as definition ids are not allowed to be qualified.

Perhaps we just need to move the check :)

peter-b replied:
@runrevmark Note that `module com.livecode.canvas` is the unique and slightly pathological case of "qualified id in definition context".

runrevmark committed:
lcb: Correct and improve identifier checks.

This patch changes the check on the module identifier to that
of a namespace. To do so it adds extra validation checks for
namespace components.

Additionally, it renames __exit in system.lcb to __Exit to avoid
a warning.

runrevmark committed:
Merge branch 'develop' into lcb-safe_foreign_handlers

Conflicts:
	tests/lcb/compiler/frontend/namespace.compilertest

runrevmark committed:
lcb: Make sure unittest module uses impl and not IMPL

runrevmark committed:
lcb: Disable check for lowercase namespace components

This patch disables the check for lowercase namespace components
as it causes a build failure due to public extensions with cased
namespaces.

runrevmark committed:
Merge branch 'develop' into lcb-safe_foreign_handlers

runrevmark committed:
lcb: Re-enable namespace case check and update extensions

This patch re-enables the namespace case check, and updates all
extensions to use a fully lower-case name. This should not have
any effect, as the names of modules is compared case-lessly
in libscript.

runrevmark committed:
lcb: Update remaining extensions to use new cased names

This patch updates the remaining extensions (compiled when running
tests) to ensure they conform to the lower case namespace
requirement.

runrevmark committed:
lcb: Update references to mixed case widgets in tests

This patch ensures that the widget tests use lowercase names
for widgets so that they work on cased filesystems (e.g. linux).

runrevmark committed:
lcb: Finish tweaking names of modules in extensions tests

This patch updates the remainder of the extension tests to use
all lowercase module names.

runrevmark committed:
lcb: Correct '__safe' in language reference

This patch corrects ____safe to __safe in the specification of the
syntax of the safe foreign handler clause.

livecodeali commented:
@livecode-vulcan review ok fe876d9

livecode-vulcan commented:
💙 review by @livecodeali ok fe876d9de25f98c8ff7f99aa0f1595fe68dc0200

livecode-vulcan commented:
😎 test success fe876d9de25f98c8ff7f99aa0f1595fe68dc0200

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1139)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1141)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1134)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1140)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1144)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1154)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1135)
